### ISSUE TITLE
Change the hash method used to store passwords

### ISSUE THREAD

**Issue by ryzom-pipeline** [ryzom/ryzomcore#206]
**[Original report](https://bitbucket.org/ryzom/ryzomcore/issue/206) by Rodolphe Breard (Bitbucket: [rodolphe_breard](https://bitbucket.org/rodolphe_breard), ).**

----------------------------------------

Currently, passwords are hashed using `crypt(3)` with the [traditional DES-based scheme](https://en.wikipedia.org/wiki/Crypt_(C)#Traditional_DES-based_scheme). This hash method was acceptable 10 years ago when Ryzom was released but is now known to be weak and [pretty fast to break](https://twitter.com/hashcat/status/160488271267364864). Therefore, storing the passwords hashed in this way is almost equivalent to storing them in clear text. Furthermore, I would like to point out that the DES-based scheme limits the password's length up to 8 characters, which is really too short. Considering [the possibility of a DB leak](http://app.ryzom.com/app_forum/index.php?page=topic/view/21036/13), the hash method should be changed to something more secured.

A simple and good enough solution is to use `crypt(3)` with a 16 characters salt prefixed by `$6$`. Such a salt would force the use of SHA-512 instead of DES. It's way more secure and allows passwords of an unlimited length. Keeping the `crypt(3)` function also has the advantage to easily allow both passwords hash methods to exist: passwords hashed with DES will still work and every new one will be hashed using SHA-512. The only downside is that the database has to be modified: `user.users.Password` must be extended from 13 to 106 characters.

I'm already working on the SHA-512 solution and it seems to work just fine. Maybe I'll be able to submit a pull request in a few days after some tests.

I heard some of you couldn't believe that crypt using DES limits the password's length or think that some magical stuff would limit it only on the C function and not the PHP one. Let me show you two examples:

```
#!php

<?php

$str = '123456';
$salt = 'ab';

foreach (str_split('7890') as $c) {
  $str = $str . $c;
  $pwd = crypt($str, $salt);
  echo 'String: ', $str, ' (', strlen($str), ' chars long): ', $pwd, PHP_EOL;
}
```

```
#!text

String: 1234567 (7 chars long): ablk9HoaAwzxk
String: 12345678 (8 chars long): ab1iBa.N.U2C6
String: 123456789 (9 chars long): ab1iBa.N.U2C6
String: 1234567890 (10 chars long): ab1iBa.N.U2C6
```

```
#!c

#define _XOPEN_SOURCE
#include <unistd.h>
#include <string.h>
#include <stdio.h>

int
main(int argc, char *argv[])
{
  char *hash;
  unsigned int l;

  if (argc < 2) {
    fprintf(stderr, "Usage: %s string_to_hash\n", argv[0]);
    return 1;
  }
  hash = crypt(argv[1], "ab");
  l = strlen(argv[1]);
  printf("String length: %u characters.\n", l);
  printf("Hash: %s\n", hash);
  return 0;
}
```

```
#!text

-> ./a.out 1234567
String length: 7 characters.
Hash: ablk9HoaAwzxk
-> ./a.out 12345678
String length: 8 characters.
Hash: ab1iBa.N.U2C6
-> ./a.out 123456789
String length: 9 characters.
Hash: ab1iBa.N.U2C6
-> ./a.out 1234567890
String length: 10 characters.
Hash: ab1iBa.N.U2C6
```

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

Putting this on high priority.

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

fixing #206 : passwords are now stored using SHA-512 instead of traditional DES

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

Not cross-platform yet

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

ref #206 : rolling back to include a crypt(3) implementation and adding sha512 support

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

* Please remove the unused link to crypt library in the CMake files
* Header "stdint.h" does not exist under older VS compilers, please include <nel/misc/types_nl.h> in top, and use uint32 etc instead of uint32_t in crypt_sha512.cpp

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

ref #206 : removing unused crypt lib reference

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

ref #206 : hacking the types so it may compile on windows

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

Merged in rodolphe_breard/ryzomcore/sha512-auth (pull request #89)

ref #206 : windows compilation stuff

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

Merged in rodolphe_breard/ryzomcore/sha512-auth (pull request #89)

ref #206 : windows compilation stuff

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

Test result:

* Account creation in AMS: OK, created with new password format
* Login in AMS, account with old password: OK
* Login in AMS, account with new password: OK
* Login using old client, account with old password: OK
* Login using updated client, account with old password: OK
* **Login using updated client, account with new password: "Error : invalid password (2004)"**

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

Log for user "sha512" logging in with password "sha512" is as follows:

2014-10-06 16:39:56 (81.244.245.84:/login/r2_login.php?cmd=login&login=sha512&password=$62dvNiNEOgmQ&clientApplication=live&cp=1&lg=en): 0:Invalid password (2004)\n

Password in database is

$6$BgKqRrdv4VZGQpFb$KSIpzmeZNj/9IkGQXfB5vg/ceWIOo5QVE3LGLqx1GsXOH.05Y/N5bzlD/cs9ZX83P83xvZWXyl4voU28zC52F/

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

Call to CCrypt::crypt(password, Salt); happens with

* sha512
* $6$BgKqRrdv4VZGQpFb

returns

* $62dvNiNEOgmQ

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

What the fuck, returning a pointer to data that's on the stack?!

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

Okay, got it working. There were multiple bugs.

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

Branch is functionally okay now, just need someone to vouch that the encryption is legit.

**Comment by ryzom-pipeline** [comment]
**Original comment by Jan Boon (Bitbucket: [Jan Boon](https://bitbucket.org/Jan Boon), ).**

----------------------------------------

Merge with sha512-auth, fix #206
